lightclient/src/platform/wasm_helpers.rs

```diff
@@ -40,87 +30,381 @@ pub fn sleep(duration: Duration) -> Delay {
-) -> Result<PlatformConnection<Stream, std::convert::Infallible>, ConnectError> {
-    // Ensure ahead of time that the multiaddress is supported.
-    let addr = match (&proto1, &proto2, &proto3) {
-        (ProtocolRef::Ip4(ip), ProtocolRef::Tcp(port), Some(ProtocolRef::Ws)) => {
-            let addr = SocketAddr::new(IpAddr::V4((*ip).into()), *port);
-            format!("ws://{}", addr.to_string())
+            WithBuffers {
+                socket,
+                error: None,
+                read_buffer: Vec::with_capacity(read_buffer_reasonable_capacity),
+                read_buffer_valid: 0,
+                read_buffer_reasonable_capacity,
+                read_closed: false,
+                write_buffers: Vec::with_capacity(64),
+                write_closed: false,
+                close_pending: false,
+                flush_pending: false,
+                read_write_now: None,
+                read_write_wake_up_after: None,
+            }
         }
-        (ProtocolRef::Ip6(ip), ProtocolRef::Tcp(port), Some(ProtocolRef::Ws)) => {
-            let addr = SocketAddr::new(IpAddr::V6((*ip).into()), *port);
-            format!("ws://{}", addr.to_string())
+
+        /// Returns an object that implements `Deref<Target = ReadWrite>`. This object can be used
+        /// to push or pull data to/from the socket.
+        ///
+        /// > **Note**: The parameter requires `Self` to be pinned for consistency with
+        /// >           [`WithBuffers::wait_read_write_again`].
+        pub fn read_write_access(
+            self: Pin<&mut Self>,
+            now: Instant,
+        ) -> Result<ReadWriteAccess, &io::Error> {
+            let this = self.project();
+
+            debug_assert!(this
+                .read_write_now
+                .as_ref()
+                .map_or(true, |old_now| *old_now <= now));
+            *this.read_write_wake_up_after = None;
+            *this.read_write_now = Some(now);
+
+            if let Some(error) = this.error.as_ref() {
+                return Err(error);
+            }
+
+            this.read_buffer.truncate(*this.read_buffer_valid);
+
+            let write_bytes_queued = this.write_buffers.iter().map(Vec::len).sum();
+
+            Ok(ReadWriteAccess {
+                read_buffer_len_before: this.read_buffer.len(),
+                write_buffers_len_before: this.write_buffers.len(),
+                read_write: read_write::ReadWrite {
+                    now,
+                    incoming_buffer: mem::take(this.read_buffer),
+                    expected_incoming_bytes: if !*this.read_closed { Some(0) } else { None },
+                    read_bytes: 0,
+                    write_bytes_queued,
+                    write_buffers: mem::take(this.write_buffers),
+                    write_bytes_queueable: if !*this.write_closed {
+                        // Limit outgoing buffer size to 128kiB.
+                        // TODO: make configurable?
```

Comment:
I think it's fine to hard-code this because it will likely to be replaced the next smoldot version

Reply:
As always, it's preferred to open issues for TODOs